lib/charset.js

```diff
@@ -0,0 +1,172 @@
+// Load modules
+
+var Boom = require('boom');
+var Hoek = require('hoek');
+
+
+// Declare internals
+
+var internals = {};
+
+
+// From https://tools.ietf.org/html/rfc7231#section-5.3.3
+// Accept-Charset: iso-8859-5, unicode-1-1;q=0.8
+
+exports.charset = function (header, preferences) {
+
+    Hoek.assert(!preferences || Array.isArray(preferences), 'Preferences must be an array');
+    var charsets = header
+        .split(',')
+        .map(internals.getParts)
+        .filter(internals.removeEmpty)
+        .sort(internals.compareByWeight);
+
+    // Tack on a default return
+
+    charsets.push({
+            weight: 0.001,
+            charset: ''
+        });
+
+    // No preferences.  Take the first non-disallowed charset
+
+    if (!preferences || preferences.length === 0) {
+        return charsets.filter(internals.removeDisallowed)[0].charset;
+    }
+
+    // Lower case all preferences
+
+    preferences = preferences.map(internals.lowerCase);
+
+    // Remove any disallowed preferences
+
+    internals.removeDisallowedPreferences(charsets, preferences);
+
+    // If charsets includes * (that isn't disallowed *;q=0) return first preference
+
+    var splatLocation = internals.findCharsetItem(charsets, '*');
+    if (splatLocation !== -1 && charsets[splatLocation].weight > 0) {
+        return preferences[0];
+    }
+
+    // Try to find the first match in the array of preferences, ignoring case
+
+    for (var i = 0, il = charsets.length; i < il; ++i) {
+        if (preferences.indexOf(charsets[i].charset.toLowerCase()) !== -1 && charsets[i].weight > 0) {
+            return charsets[i].charset;
+        }
+    }
+
+    return '';
+};
+
+
+exports.charsets = function (header) {
+
+    if (header === undefined || typeof header !== 'string') {
+        return [];
+    }
+
+    header = header.toLowerCase();
+
+    return header
+        .split(',')
+        .map(internals.getParts)
+        .filter(internals.removeEmptyAndDisallowed)
+        .sort(internals.compareByWeight)
+        .map(internals.partToCharset);
```

Comment:
Might be worth trying this out with a for loop and a single sort after for performance. My experience is that this much functional JS tends to be very slow. You can look into it later in another PR.